lib/src/geometry/polygon.dart

```diff
@@ -0,0 +1,156 @@
+import 'dart:math';
+import 'dart:ui';
+
+import '../../components.dart';
+import '../extensions/rect.dart';
+import '../extensions/vector2.dart';
+import 'shape.dart';
+
+class Polygon extends Shape {
+  final List<Vector2> definition;
+
+  /// With this constructor you create your [Polygon] from positions in your
+  /// intended space. It will automatically calculate the [size] and center
+  /// ([position]) of the Polygon.
+  factory Polygon(
+    List<Vector2> points, {
+    double angle = 0,
+  }) {
+    final center = points.fold<Vector2>(
+          Vector2.zero(),
+          (sum, v) => sum + v,
+        ) /
+        points.length.toDouble();
+    final bottomRight = points.fold<Vector2>(
+      Vector2.zero(),
+      (bottomRight, v) {
+        return Vector2(
+          max(bottomRight.x, v.x),
+          max(bottomRight.y, v.y),
+        );
+      },
+    );
+    final halfSize = bottomRight - center;
+    final definition =
+        points.map<Vector2>((v) => (v - center)..divide(halfSize)).toList();
+    return Polygon.fromDefinition(
+      definition,
+      position: center,
+      size: halfSize * 2,
+      angle: angle,
+    );
+  }
+
+  /// With this constructor you define the [Polygon] from the center of and with
+  /// percentages of the size of the shape.
+  /// Example: [[1.0, 0.0], [0.0, 1.0], [-1.0, 0.0], [0.0, -1.0]]
+  /// This will form a diamond shape within the bounding size box.
+  /// NOTE: Always define your shape is a clockwise fashion
+  Polygon.fromDefinition(
+    this.definition, {
+    Vector2 position,
+    Vector2 size,
+    double angle,
+  }) : super(position: position, size: size, angle: angle ?? 0);
+
+  final _cachedScaledShape = ShapeCache<Iterable<Vector2>>();
+
+  /// Gives back the shape vectors multiplied by the size
+  Iterable<Vector2> get scaled {
+    if (!_cachedScaledShape.isCacheValid(<dynamic>[size])) {
```

Comment:
I added some comments here https://github.com/flame-engine/flame/pull/633/files?short_path=b9b2088#diff-a0e482c67889baed9504d4fccebd82f03bfe0c45432255ab846f6e8bfd5478ebR84

on how to avoid us needing the <dynamic> cast